tempodb/blocklist/poller.go

```diff
@@ -148,44 +150,73 @@ func (p *Poller) Do(previous *List) (PerTenant, PerTenantCompacted, error) {
 		return nil, nil, err
 	}
 
-	blocklist := PerTenant{}
-	compactedBlocklist := PerTenantCompacted{}
+	var (
+		wg  = boundedwaitgroup.New(p.cfg.TenantPollConcurrency)
+		mtx = sync.Mutex{}
 
-	consecutiveErrors := 0
+		blocklist          = PerTenant{}
+		compactedBlocklist = PerTenantCompacted{}
 
-	for _, tenantID := range tenants {
-		newBlockList, newCompactedBlockList, err := p.pollTenantAndCreateIndex(ctx, tenantID, previous)
-		if err != nil {
-			level.Error(p.logger).Log("msg", "failed to poll or create index for tenant", "tenant", tenantID, "err", err)
-			consecutiveErrors++
-			if consecutiveErrors > p.cfg.TolerateConsecutiveErrors {
-				level.Error(p.logger).Log("msg", "exiting polling loop early because too many errors", "errCount", consecutiveErrors)
-				return nil, nil, err
-			}
+		finalErr atomic.Error
+		pollErrs = make([]error, len(tenants))
```

Comment:
instead of tracking errors with this slice i think we can do something like this in the go func():

```
go func(tenantID string) {
			defer wg.Done()
			var err error 
			errLefts := p.cfg.TolerateConsecutiveErrors

                         for errsLeft > 0 {
   			     newBlockList, newCompactedBlockList, err := p.pollTenantAndCreateIndex(ctx, tenantID, previous)
   			     if err == nil {
   			        break
   			    }
   			    
   			    // log error and errsLeft count
   			    errsLeft--
   			}
   			
   			if err != nil {
   			   // fatal err!
   			}
   			...
```
			
I like that this keeps the bookkeeping of errors local to the function and doesn't require tracking it with a slice across all tenants.

Reply:
Hmm, interesting.  So this would move the error toleration from global to a per-tenant, which sounds pretty nice.  I think this means we could fail all but one tenant by default and still call the poll a success.  I'm not sure how I feel about that.

Since this used to be sequential, checking consecutive error checking here made sense, but I don't know what to make of it when running concurrently.  Does this just need to become a "total errors while polling" config option, rather than a "consequetive" toleration?  Then we could do like you say, but instead of `errLefts` being defined in the routine, we move it to the main tenant loop and do something like you've got there.

Reply:
I added a commit for this, and an additional config for the new behavior.  I think we don't want a single failing tenant to bring down the entire poll by default.  This sort of merges old behavior with new behavior, and includes the encapsulation mentioned above.  I think it looks pretty tidy at this point, but let me know your thoughts.  Let's sync if we need to.